File: app/src/main/java/com/battlelancer/seriesguide/jobs/episodes/EpisodeWatchedJob.java
package com.battlelancer.seriesguide.jobs.episodes;

import android.content.Context;
import androidx.annotation.NonNull;
import com.battlelancer.seriesguide.R;
import com.battlelancer.seriesguide.appwidget.ListWidgetProvider;
import com.battlelancer.seriesguide.provider.SgActivityHelper;
import com.battlelancer.seriesguide.provider.SgEpisode2Helper;
import com.battlelancer.seriesguide.provider.SgRoomDatabase;
import com.battlelancer.seriesguide.ui.episodes.EpisodeFlags;
import com.battlelancer.seriesguide.ui.episodes.EpisodeTools;
import com.battlelancer.seriesguide.util.TextTools;

public class EpisodeWatchedJob extends EpisodeBaseJob {

    public EpisodeWatchedJob(long episodeId, int episodeFlags) {
        super(episodeId, episodeFlags, JobAction.EPISODE_WATCHED_FLAG);
    }

    private long getLastWatchedEpisodeId(Context context) {
        if (!EpisodeTools.isUnwatched(getFlagValue())) {
            // watched or skipped episode
            return episodeId;
        } else {
            // changed episode to not watched
            long lastWatchedId = -1; // don't change last watched episode by default

            // if modified episode is identical to last watched one (e.g. was just watched),
            // find an appropriate last watched episode
            SgRoomDatabase database = SgRoomDatabase.getInstance(context);
            long lastWatchedEpisodeId = database.sgShow2Helper()
                    .getShowLastWatchedEpisodeId(getShowId());
            // identical to last watched episode?
            if (episodeId == lastWatchedEpisodeId) {
                if (getEpisode().getSeason() == 0) {
                    // keep last watched (= this episode) if we got a special
                    return -1;
                }
                lastWatchedId = 0; // re-set if we don't find one

                // get newest watched before this one
                long previousWatchedEpisodeId = database.sgEpisode2Helper()
                        .getPreviousWatchedEpisodeOfShow(getShowId(), getEpisode().getSeason(),
                                getEpisode().getEpisodenumber());
                if (previousWatchedEpisodeId > 0) {
                    lastWatchedId = previousWatchedEpisodeId;
                }
            }

            return lastWatchedId;
        }
    }

    @Override
    public boolean applyLocalChanges(Context context, boolean requiresNetworkJob) {
        if (!super.applyLocalChanges(context, requiresNetworkJob)) {
            return false;
        }

        // set a new last watched episode
        // set last watched time to now if marking as watched or skipped
        boolean unwatched = EpisodeTools.isUnwatched(getFlagValue());
        updateLastWatched(context, getLastWatchedEpisodeId(context), !unwatched);

        if (EpisodeTools.isWatched(getFlagValue())) {
            // create activity entry for watched episode
            SgActivityHelper.addActivity(context, episodeId, getShowId());
        } else if (unwatched) {
            // remove any previous activity entries for this episode
            // use case: user accidentally toggled watched flag
            SgActivityHelper.removeActivity(context, episodeId);
        }

        ListWidgetProvider.notifyDataChanged(context);

        return true;
    }

    @Override
    protected boolean applyDatabaseChanges(@NonNull Context context) {
        SgEpisode2Helper episodeHelper = SgRoomDatabase.getInstance(context).sgEpisode2Helper();
        int flagValue = getFlagValue();

        int rowsUpdated;
        switch (flagValue) {
            case EpisodeFlags.SKIPPED:
                rowsUpdated = episodeHelper.setSkipped(episodeId);
                break;
            case EpisodeFlags.WATCHED:
                rowsUpdated = episodeHelper.setWatchedAndAddPlay(episodeId);
                break;
            case EpisodeFlags.UNWATCHED:
                rowsUpdated = episodeHelper.setNotWatchedAndRemovePlays(episodeId);
                break;
            default:
                throw new IllegalArgumentException("Flag value not supported");
        }

        return rowsUpdated == 1;
    }

    /**
     * Note: this should mirror the planned database changes in {@link #applyDatabaseChanges(Context)}.
     */
    @Override
    protected int getPlaysForNetworkJob(int plays) {
        switch (getFlagValue()) {
            case EpisodeFlags.SKIPPED:
                return plays;
            case EpisodeFlags.WATCHED:
                return plays + 1;
            case EpisodeFlags.UNWATCHED:
                return 0;
            default:
                throw new IllegalArgumentException("Flag value not supported");
        }
    }

    @NonNull
    @Override
    public String getConfirmationText(Context context) {
        int actionResId;
        int flagValue = getFlagValue();
        if (EpisodeTools.isSkipped(flagValue)) {
            actionResId = R.string.action_skip;
        } else if (EpisodeTools.isWatched(flagValue)) {
            actionResId = R.string.action_watched;
        } else {
            actionResId = R.string.action_unwatched;
        }
        // format like '6x42 · Set watched'
        String number = TextTools.getEpisodeNumber(context, getEpisode().getSeason(),
                getEpisode().getEpisodenumber());
        return TextTools.dotSeparate(number, context.getString(actionResId));
    }
}

File: app/src/main/java/com/battlelancer/seriesguide/jobs/episodes/EpisodeWatchedJob.kt
// SPDX-License-Identifier: Apache-2.0
// Copyright 2017, 2018, 2020, 2021, 2023, 2024 Uwe Trottmann
package com.battlelancer.seriesguide.jobs.episodes

import android.content.Context
import com.battlelancer.seriesguide.appwidget.ListWidgetProvider
import com.battlelancer.seriesguide.provider.SgRoomDatabase
import com.battlelancer.seriesguide.shows.database.SgEpisode2Numbers
import com.battlelancer.seriesguide.shows.episodes.EpisodeFlags
import com.battlelancer.seriesguide.shows.episodes.EpisodeTools

class EpisodeWatchedJob(
    episodeId: Long,
    episodeFlags: Int
) : EpisodeBaseJob(episodeId, episodeFlags, JobAction.EPISODE_WATCHED_FLAG) {

    private fun getLastWatchedEpisodeId(context: Context): Long {
        return if (!EpisodeTools.isUnwatched(flagValue)) {
            // watched or skipped episode
            episodeId
        } else {
            // changed episode to not watched
            var lastWatchedId: Long = -1 // don't change last watched episode by default

            // if modified episode is identical to last watched one (e.g. was just watched),
            // find an appropriate last watched episode
            val database = SgRoomDatabase.getInstance(context)
            val lastWatchedEpisodeId = database.sgShow2Helper()
                .getShowLastWatchedEpisodeId(showId)
            // identical to last watched episode?
            if (episodeId == lastWatchedEpisodeId) {
                if (episode.season == 0) {
                    // keep last watched (= this episode) if we got a special
                    return -1
                }
                lastWatchedId = 0 // re-set if we don't find one

                // get newest watched before this one
                val previousWatchedEpisodeId = database.sgEpisode2Helper()
                    .getPreviousWatchedEpisodeOfShow(
                        showId, episode.season,
                        episode.episodenumber
                    )
                if (previousWatchedEpisodeId > 0) {
                    lastWatchedId = previousWatchedEpisodeId
                }
            }
            lastWatchedId
        }
    }

    override fun applyDatabaseChanges(
        context: Context,
        episodes: List<SgEpisode2Numbers>
    ): Boolean {
        val episodeHelper = SgRoomDatabase.getInstance(context).sgEpisode2Helper()
        val flagValue = flagValue
        val rowsUpdated: Int = when (flagValue) {
            EpisodeFlags.SKIPPED -> episodeHelper.setSkipped(episodeId)
            EpisodeFlags.WATCHED -> episodeHelper.setWatchedAndAddPlay(episodeId)
            EpisodeFlags.UNWATCHED -> episodeHelper.setNotWatchedAndRemovePlays(episodeId)
            else -> throw IllegalArgumentException("Flag value not supported")
        }
        val isSuccessful = rowsUpdated == 1

        if (isSuccessful) {
            // set a new last watched episode
            // set last watched time to now if marking as watched or skipped
            val unwatched = EpisodeTools.isUnwatched(flagValue)
            updateLastWatched(context, getLastWatchedEpisodeId(context), !unwatched)

            // Add or remove activity entry
            updateActivity(context, episodes)

            ListWidgetProvider.notifyDataChanged(context)
        }
        return isSuccessful
    }

    /**
     * Note: this should mirror the planned database changes in [applyDatabaseChanges].
     */
    override fun getPlaysForNetworkJob(plays: Int): Int {
        return when (flagValue) {
            EpisodeFlags.SKIPPED -> plays
            EpisodeFlags.WATCHED -> plays + 1
            EpisodeFlags.UNWATCHED -> 0
            else -> throw IllegalArgumentException("Flag value not supported")
        }
    }
}
